src/main/scala/org/broadinstitute/dsde/firecloud/service/LibraryService.scala

```diff
@@ -119,27 +123,28 @@ class LibraryService (protected val argUserInfo: UserInfo, val rawlsDAO: RawlsDA
   }
 
   def publishDocument(ws: Workspace): Unit = {
-    searchDAO.indexDocument(indexableDocument(ws))
+    indexableDocument(ws, ontologyDAO) map { searchDAO.indexDocument }
   }
 
   def removeDocument(ws: Workspace): Unit = {
     searchDAO.deleteDocument(ws.workspaceId)
   }
 
   def indexAll: Future[PerRequestMessage] = {
-    rawlsDAO.getAllLibraryPublishedWorkspaces map {workspaces =>
+    rawlsDAO.getAllLibraryPublishedWorkspaces flatMap { workspaces: Seq[Workspace] =>
       searchDAO.recreateIndex()
       if (workspaces.isEmpty)
-        RequestComplete(NoContent)
+        Future(RequestComplete(NoContent))
       else {
-        val toIndex:Seq[Document] = workspaces.map {workspace => indexableDocument(workspace)}
-        val bi = searchDAO.bulkIndex(toIndex)
-        val statusCode = if (bi.hasFailures) {
-          InternalServerError
-        } else {
-          OK
+        val toIndex: Future[Seq[Document]] = Future.sequence(workspaces map { indexableDocument(_, ontologyDAO) })
```

Comment:
And to add to that answer, the ontology lookups to ES have been pretty zippy through all of the testing I've done. If we run into issues here, I'd be worried.